test/e2e/specs/editor/blocks/list.spec.js

```diff
@@ -175,13 +175,16 @@ test.describe( 'List (@firefox)', () => {
 		await expect.poll( editor.getEditedPostContent ).toBe( '' );
 	} );
 
-	test( 'should not undo asterisk transform with backspace after selection change (-firefox)', async ( {
+	test( 'should not undo asterisk transform with backspace after selection change', async ( {
 		editor,
 		page,
 	} ) => {
 		await page.click( 'role=button[name="Add default block"i]' );
 		await page.keyboard.type( '* ' );
 		await expect( page.locator( '[data-type="core/list"]' ) ).toBeVisible();
+		// Wait until the automatic change is marked as "final", which is done
+		// with an idle callback, see __unstableMarkAutomaticChange.
+		await page.evaluate( () => new Promise( window.requestIdleCallback ) );
```

Comment:
This is not something we want to put in e2e tests, since it's implementation details and not what end users could interact with. It can potentially also be an issue for end users if they press the `ArrowUp` key fast enough, right? Ideally, we should fix it in the code so that it can be deterministic. If that's not possible at the time then we should think along the line with retrying and waiting for the results, something like this:

```js
await expect( async () => {
	await page.keyboard.press( 'ArrowUp' );
	expect(
		await page.evaluate( () =>
			window.wp.data.select( 'core/block-editor' ).getSelectedBlock()
		)
	).toMatchObject( { name: 'core/list' } );
} ).toPass();
```

Reply:
It's not really an issue for a user: if selection changes too quickly, it's just that it will still be possible to undo the automatic change. We have a few of these lines throughout the e2e test where we have to wait for selection. We also have undo levels for rich text after 1 second of inactivity. How would we write an e2e test for something like that? Let's look at these all together separately.

Reply:
(I'd rather improve test stability for trunk right now.)

Reply:
Please follow up with a PR to fix this then.

> We have a few of these lines throughout the e2e test where we have to wait for selection.

And this is the anti-pattern that I'm recommending against. 

> We also have undo levels for rich text after 1 second of inactivity. How would we write an e2e test for something like that?

Some of these can be mocked to reduce the time we have to wait. If that's hard-coded then we can always fallback to the pattern I mentioned above: wait for feedback that's accessible to the end users.

Reply:
I think we should avoid mocking entirely in e2e tests if possible otherwise it's not e2e testing anymore. If the user has to wait, let's make the test wait as well.

Reply:
Agreed. But sometimes it's inevitable, like caching and nonce tests, or just isolating for more stable tests. Waiting is okay, we just need to wait for [user-visible behaviors](https://playwright.dev/docs/best-practices#test-user-visible-behavior) instead.

Reply:
Hi @ellatrix, friendly reminder that this exists ❤️ . Feel free to request my reviews in the PR and we can discuss alternatives there more!

Reply:
I'm not really sure what you're asking here. I added the best solution to my knowledge in this PR. What does the code above do? It just keeps pressing up until the test passes? To me that doesn't feel like an e2e test?